File: samples/fpc/architecture_comparison.pas
{
  architecture_comparison.pas — one client library, two ways into the engine.

  The fbintf twin of ../cpp/architecture_comparison.cpp, companion to
  ../../architecture-comparison.md.  The comparison table's most unusual
  row is Firebird's: "client-server AND embedded, same engine".  One
  Pascal binary, linked only against libfbclient through fbintf, proves
  it: attach to localhost:employee and the Y-valve picks the Remote
  provider; attach to a plain local path and the Y-valve loads the Engine
  provider INTO THIS PROCESS - no server involved.  For each attachment
  the engine itself reports where it runs: ENGINE_VERSION,
  NETWORK_PROTOCOL, and MON$SERVER_PID versus our own pid.

  Build & run (see ../README.md):
      make -C samples/fpc bin/architecture_comparison
      samples/fpc/bin/architecture_comparison
}
program architecture_comparison;

{$mode delphi}
{$codepage UTF8}
{$interfaces COM}

uses SysUtils, BaseUnix, IB, FBHandsOn;

{The embedded engine needs to find its root (plugins, security db).}
function setenv(name, value: PAnsiChar; overwrite: longint): longint;
  cdecl; external 'c';

procedure Inspect(const banner, conn: AnsiString; createIfAbsent: boolean);
var Att: IAttachment;
    Tr: ITransaction;
    R: IResultset;
    protocol, serverPid, samePid: AnsiString;
begin
  if createIfAbsent then
    Att := AttachOrCreate(conn)
  else
    Att := FirebirdAPI.OpenDatabase(conn, DefaultDPB);

  Tr := Att.StartTransaction([isc_tpb_read, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  R := Att.OpenCursorAtStart(Tr,
    'select rdb$get_context(''SYSTEM'', ''ENGINE_VERSION''), ' +
    '       rdb$get_context(''SYSTEM'', ''NETWORK_PROTOCOL''), ' +
    '       a.mon$server_pid ' +
    'from mon$attachments a ' +
    'where a.mon$attachment_id = current_connection');

  if R[1].IsNull then protocol := '(null)' else protocol := R[1].AsString;
  serverPid := R[2].AsString;
  if serverPid = IntToStr(FpGetPid) then
    samePid := ' -- the engine runs IN this process'
  else
    samePid := '';

  writeln(banner);
  writeln('    connection string : ', conn);
  writeln('    ENGINE_VERSION    : ', R[0].AsString);
  writeln('    NETWORK_PROTOCOL  : ', protocol);
  writeln('    MON$SERVER_PID    : ', serverPid, '   (this process is pid ',
    FpGetPid, samePid, ')');
  Tr.Commit;
end;

var remoteDb, embeddedDb: AnsiString;

begin
  remoteDb := 'localhost:employee';
  if ParamCount >= 1 then remoteDb := ParamStr(1);
  embeddedDb := '/tmp/fbhandson/arch_embedded_fpc.fdb';
  if ParamCount >= 2 then embeddedDb := ParamStr(2);

  setenv('FIREBIRD', '/opt/firebird', 0);

  writeln('One libfbclient behind fbintf, two providers behind the Y-valve.');
  writeln;
  Inspect('[1] Remote provider (client-server):', remoteDb, false);
  writeln;
  Inspect('[2] Engine provider (embedded, no server):', embeddedDb, true);
  writeln;
  writeln('done.');
end.

File: samples/cpp/architecture_comparison.cpp
/*
 *  architecture_comparison.cpp — one client library, two ways into the engine.
 *
 *  Companion sample for ../../architecture-comparison.md.  The comparison
 *  table's most unusual row is Firebird's: "client-server AND embedded,
 *  same engine".  This program proves it with a single binary linked only
 *  against libfbclient:
 *
 *    1. attach inet://localhost/employee      — the Y-valve picks the Remote
 *       provider; SQL travels over the wire to the server process.
 *    2. attach /tmp/fbhandson/arch_embedded.fdb (a plain local path; created
 *       on first run) — the Y-valve loads the Engine provider INTO THIS
 *       PROCESS; the same libfbclient is now a complete database engine and
 *       no server is involved at all.
 *
 *  For each attachment we ask the engine itself where we are:
 *    - ENGINE_VERSION            — same engine code both ways
 *    - NETWORK_PROTOCOL          — TCPv4/TCPv6 remotely, NULL embedded
 *    - MON$SERVER_PID vs getpid()— the server's process remotely,
 *                                  OUR OWN pid when embedded
 */

#include "fb_sample.h"
#include <unistd.h>

using namespace fbsample;

static void inspect(const char* label, const char* database, bool create)
{
	Db db;
	if (create)
		db.attachOrCreate(database);
	else
		db.attach(database);

	ITransaction* tra = db.start();

	Db::Table t = db.query(tra,
		"select rdb$get_context('SYSTEM', 'ENGINE_VERSION'), "
		"       rdb$get_context('SYSTEM', 'NETWORK_PROTOCOL'), "
		"       a.mon$server_pid "
		"from mon$attachments a "
		"where a.mon$attachment_id = current_connection");

	const std::string& version = t.rows[0][0];
	const std::string& protocol = t.rows[0][1];
	const std::string& serverPid = t.rows[0][2];

	printf("%s\n", label);
	printf("    connection string : %s\n", database);
	printf("    ENGINE_VERSION    : %s\n", version.c_str());
	printf("    NETWORK_PROTOCOL  : %s\n", protocol.c_str());
	printf("    MON$SERVER_PID    : %s   (this process is pid %d%s)\n",
		serverPid.c_str(), getpid(),
		serverPid == std::to_string(getpid())
			? " -- the engine runs IN this process" : "");

	tra->commit(&db.status);
}

int main(int argc, char** argv)
{
	const char* remoteDb = argOrDefault(argc, argv, 1,
		"inet://localhost/employee");
	const char* embeddedDb = argOrDefault(argc, argv, 2,
		"/tmp/fbhandson/arch_embedded.fdb");

	try
	{
		printf("One libfbclient, two providers behind the Y-valve.\n\n");
		inspect("[1] Remote provider (client-server):", remoteDb, false);
		printf("\n");
		inspect("[2] Engine provider (embedded, no server):", embeddedDb, true);
		printf("\ndone.\n");
		return 0;
	}
	catch (const FbException& e)
	{
		return report(e);
	}
}
